impeller/entity/contents/atlas_contents.cc

```diff
@@ -121,31 +189,103 @@ bool AtlasContents::Render(const ContentContext& renderer,
+  frag_info.texture_sampler_y_coord_scale = texture->GetYCoordScale();
   frag_info.alpha = alpha_;
 
-  Command cmd;
-  cmd.label = "DrawAtlas";
-  cmd.pipeline =
-      renderer.GetAtlasPipeline(OptionsFromPassAndEntity(pass, entity));
+  auto options = OptionsFromPassAndEntity(pass, entity);
+  cmd.pipeline = renderer.GetTexturePipeline(options);
   cmd.stencil_reference = entity.GetStencilDepth();
   cmd.BindVertices(vertex_builder.CreateVertexBuffer(host_buffer));
   VS::BindVertInfo(cmd, host_buffer.EmplaceUniform(vert_info));
   FS::BindFragInfo(cmd, host_buffer.EmplaceUniform(frag_info));
-  FS::BindTextureSampler(cmd, texture_,
+  FS::BindTextureSampler(cmd, texture,
                          renderer.GetContext()->GetSamplerLibrary()->GetSampler(
-                             sampler_descriptor_));
-  pass.AddCommand(std::move(cmd));
+                             parent_.GetSamplerDescriptor()));
+  return pass.AddCommand(std::move(cmd));
+}
+
+// AtlasColorContents
+// ---------------------------------------------------------
+
+AtlasColorContents::AtlasColorContents(const AtlasContents& parent)
+    : parent_(parent) {}
 
-  return true;
+AtlasColorContents::~AtlasColorContents() {}
+
+std::optional<Rect> AtlasColorContents::GetCoverage(
+    const Entity& entity) const {
+  return coverage_.TransformBounds(entity.GetTransformation());
+}
+
+void AtlasColorContents::SetAlpha(Scalar alpha) {
+  alpha_ = alpha;
+}
+
+void AtlasColorContents::SetCoverage(Rect coverage) {
+  coverage_ = coverage;
+}
+
+bool AtlasColorContents::Render(const ContentContext& renderer,
+                                const Entity& entity,
+                                RenderPass& pass) const {
+  using VS = GeometryColorPipeline::VertexShader;
+  using FS = GeometryColorPipeline::FragmentShader;
+
+  auto texture_coords = parent_.GetTextureCoordinates();
+  auto transforms = parent_.GetTransforms();
+  auto colors = parent_.GetColors();
+
+  VertexBufferBuilder<VS::PerVertexData> vertex_builder;
+  vertex_builder.Reserve(texture_coords.size() * 6);
+  constexpr size_t indices[6] = {0, 1, 2, 1, 2, 3};
+  for (size_t i = 0; i < texture_coords.size(); i++) {
+    auto sample_rect = texture_coords[i];
+    auto matrix = transforms[i];
+    auto transformed_points =
+        Rect::MakeSize(sample_rect.size).GetTransformedPoints(matrix);
+
+    for (size_t j = 0; j < 6; j++) {
+      VS::PerVertexData data;
+      data.position = transformed_points[indices[j]];
+      data.color = colors[i].Premultiply();
+      vertex_builder.AppendVertex(data);
+    }
+  }
+
+  if (!vertex_builder.HasVertices()) {
+    return true;
+  }
+
+  Command cmd;
+  cmd.label = "AtlasColors";
+
+  auto& host_buffer = pass.GetTransientsBuffer();
+
+  VS::VertInfo vert_info;
+  vert_info.mvp = Matrix::MakeOrthographic(pass.GetRenderTargetSize()) *
+                  entity.GetTransformation();
+
+  FS::FragInfo frag_info;
+  frag_info.alpha = alpha_;
```

Comment:
Ok, so looking over this, I _think_ the destination color is getting double premultiplied. First it's getting multiplied here with the `GeometryColorPipeline` , and then the same value is getting multiplied into the destination via the new alpha in the blend filter.

I think the better place to keep the `SetAlpha` business is here in this contents, and the new `SetAlpha` stuff can be removed from the blend filter.

If it still doesn't look like Skia after fixing the double premul, then I'd wonder if this alpha is supposed to be deferred to the final blend as opposed to impacting the destination color of the vertex color blend (I suspect you may have worked this out already, though).

Reply:
That is what I thought, but I don't set the alpha here at all if I'm blending - its only used for the src or dst case. Double checking this